# T02
## Marianne Trujillo Altamirano
### *Exp. 740694*

# Tarea - Programación lineal

### Diseño de la Dieta Óptima

Se quiere producir comida para gatos de la manera más barata, no obstante se debe también asegurar que se cumplan los datos requeridos de analisis nutricional. Por lo que se quiere variar la cantidad de cada ingrediente para cumplir con los estandares nutricionales. Los requisitos que se tienen es que en $100$ gramos, se deben tener **por lo menos** $8$ gramos de proteína y $6$ gramos de grasa. Así mismo, no se debe tener más de $2$ gramos de fibra y $0.4$ gramos de sal.  

Los datos nutricionales se pueden obtener de la siguiente tabla:

Ingrediente|Proteína|Grasa|Fibra|Sal
:----|----|----|----|----
Pollo|  10.0%|08.0%|00.1%|00.2%
Carne|  20.0%|10.0%|00.5%|00.5%
Cordero|15.0%|11.0%|00.5%|00.7%
Arroz|  00.0%|01.0%|10.0%|00.2%
Trigo|  04.0%|01.0%|15.0%|00.8%
Gel|    00.0%|00.0%|00.0%|00.0%

Los costos de cada producto son:

Ingrediente|Costo por gramo
:----|----
Pollo|$\$$0.013
Carne|$\$$0.008
Cordero|$\$$0.010
Arroz|$\$$0.002
Trigo|$\$$0.005
Gel|$\$$0.001    

Lo que se busca optimizar en este caso es la cantidad de productos que se debe utilizar en la comida de gato, minimizando el costo total. Para simplificar la notación use las siguientes variables: 

+ $x_1:$ Gramos de pollo  
+ $x_2:$ Gramos de carne  
+ $x_3:$ Gramos de cordero  
+ $x_4:$ Gramos de arroz  
+ $x_5:$ Gramos de trigo  
+ $x_6:$ Gramos de gel

La tarea consiste en plantear el problemade programación lineal que permita satisfacer las necesidades alimenticias del gato al tiempo que minimice el costo total y resolverlo con `linprog`.

### **1. Criterio de optimiación**

Minimizar el costo total de producción

### **2. Variables que afectan mi criterio**

Cantidad de ingredientes a usar

### **3. Función de costo**

$$0.013x_1+0.008x_2+0.010x_3+0.002x_4+0.005x_5+0.001x_6$$

### **4. Restricciones**

Gramos totales
$$x_1+x_2+x_3+x_4+x_5+x_6=100$$
Proteína
$$0.1x_1+0.2x_2+0.15x_3+0.04x_5 \geq 8$$
Grasa
$$0.08x_1+0.1x_2+0.11x_3+0.01x_4+0.01x_5 \geq 6$$
Fibra
$$0.001x_1+0.005x_2+0.005x_3+0.1x_4+0.15x_5 \leq 2$$
Sal
$$0.002x_1+0.005x_2+0.007x_3+0.002x_4+0.008x_5 \leq 0.4$$

In [16]:
import scipy.optimize as opt
import numpy as np

c = np.array([0.013,
              0.008,
              0.010,
              0.002,
              0.005,
              0.001])

Aeq = np.array ([
               [1, 1, 1, 1, 1, 1]
               ])

beq = np.array ([100])


Aub = np.array([
               [-0.1, -0.2, -0.15, 0, -0.04,0],
               [-0.08, -0.1, -0.11, -0.01, -0.01, 0],
               [0.001, 0.005, 0.005, 0.10, 0.15, 0],
               [0.002, 0.005, 0.007, 0.002, 0.008, 0]
               ])

bub = np.array([-8,
                -6,
                 2,
                 0.4])



opt.linprog(c, Aub, bub, Aeq, beq)

           con: array([0.])
 crossover_nit: 0
         eqlin:  marginals: array([0.001])
  residual: array([0.])
           fun: 0.52
       ineqlin:  marginals: array([-0.  , -0.07, -0.  , -0.  ])
  residual: array([4. , 0. , 1.7, 0.1])
         lower:  marginals: array([0.0064, 0.    , 0.0013, 0.0003, 0.0033, 0.    ])
  residual: array([ 0., 60.,  0.,  0.,  0., 40.])
       message: 'Optimization terminated successfully. (HiGHS Status 7: Optimal)'
           nit: 2
         slack: array([4. , 0. , 1.7, 0.1])
        status: 0
       success: True
         upper:  marginals: array([0., 0., 0., 0., 0., 0.])
  residual: array([inf, inf, inf, inf, inf, inf])
             x: array([ 0., 60.,  0.,  0.,  0., 40.])

**Análisis de resultados**
La cantidad de cada ingrediente que se debe usar para minimizar el costo total es de:
+ $x_1:$ Gramos de pollo (0)  
+ $x_2:$ Gramos de carne (60 gr) 
+ $x_3:$ Gramos de cordero (0)
+ $x_4:$ Gramos de arroz  (0)
+ $x_5:$ Gramos de trigo  (0)
+ $x_6:$ Gramos de gel (40)

### Problema de distribución de energía eléctrica

La Comisión Federal de Electricidad **(CFE)** dispone de cuatro plantas de generación para satisfacer la demanda diaria eléctrica en cuatro ciudades, Guadalajara, León y Morelia. Las plantas $1$, $2$ y $3$ pueden satisfacer $80$, $40$ y $60$ millones de KW al día respectivamente. Las necesidades de las ciudades de Guadalajara, León y Morelia son de $70$, $40$ y $70$ millones de Kw al día respectivamente. 


Los costos asociados al envío de suministro energético por cada millón de Kw entre cada planta y cada ciudad son los registrados en la siguiente tabla. 

-|Guadalajara|León|Morelia
:----|----|----|----
Planta 1|5|2|7
Planta 2|3|6|6
Planta 3|6|1|2

Y por último, las restricciones del problema, van a estar dadas por las capacidades de oferta y demanda de cada planta (en millones de KW) y cada ciudad.

Para simplificar la notación use las siguientes variables:

+ $x_1$: Kw (en millones) distribuidos de la Planta 1 a Guadalajara
+ $x_2$: Kw (en millones) distribuidos de la Planta 1 a León
+ $x_3$: Kw (en millones) distribuidos de la Planta 1 a Morelia
+ $x_4$: Kw (en millones) distribuidos de la Planta 2 a Guadalajara
+ $x_5$: Kw (en millones) distribuidos de la Planta 2 a León
+ $x_6$: Kw (en millones) distribuidos de la Planta 2 a Morelia
+ $x_7$: Kw (en millones) distribuidos de la Planta 3 a Guadalajara
+ $x_8$: Kw (en millones) distribuidos de la Planta 3 a León
+ $x_9$: Kw (en millones) distribuidos de la Planta 3 a Morelia

La tarea consiste en plantear el problema de programación lineal que permita satisfacer las necesidades de todas las ciudades al tiempo que minimice los costos asociados a la distribución y resolverlo con `linprog`.

### **1. Criterio de optimiación**

Minimizar los costos asociados a la producción

### **2. Variables que afectan mi criterio**

Capacidad de oferta y demanda de cada planta.

### **3. Función de costo**


$$\min_{x_1,\dots,x_6}2x_{AL} + 7x_{AM} + 5x_{AG} + 6x_{BL} + 6x_{BM} + 3x_{BG} + x_{CL} + 2x_{CM} + 6x_{CG}$$

### **4. Restricciones**

Restricciones de la producción:

- $x_{AL} + x_{AM} + x_{AG} \leq 80$
- $x_{BL} + x_{BM} + x_{BG} \leq 40$
- $x_{CL} + x_{CM} + x_{CG} \leq 60$


Restricciones del consumo:

- $x_{AL} + x_{BL} + x_{CL} \geq 40$
- $x_{AM} + x_{BM} + x_{CM} \geq 70$
- $x_{AG} + x_{BG} + x_{CG} \geq 70$






In [8]:
import scipy.optimize as opt
import numpy as np

c = np.array([2,
              7,
              5,
              6,
              6,
              3,
              1,
              2,
              6
              ])


Aub = np.array([[1, 1, 1, 0, 0, 0, 0, 0, 0],
                [0, 0, 0, 1, 1, 1, 0, 0, 0],
                [0, 0, 0, 0, 0, 0, 1, 1, 1],
              ])

bub = np.array([80,
                40,
                60
               ])

Aeq = np.array([[1, 0, 0, 1, 0, 0, 1, 0, 0],
                [0, 1, 0, 0, 1, 0, 0, 1, 0],
                [0, 0, 1, 0, 0, 1, 0, 0, 1]
               ])

beq = np.array([40,
                70,
                70
                ])


opt.linprog(c, Aub, bub, Aeq, beq)

           con: array([0., 0., 0.])
 crossover_nit: 0
         eqlin:  marginals: array([2., 7., 5.])
  residual: array([0., 0., 0.])
           fun: 540.0
       ineqlin:  marginals: array([-0., -2., -5.])
  residual: array([0., 0., 0.])
         lower:  marginals: array([0., 0., 0., 6., 1., 0., 4., 0., 6.])
  residual: array([40., 10., 30.,  0.,  0., 40.,  0., 60.,  0.])
       message: 'Optimization terminated successfully. (HiGHS Status 7: Optimal)'
           nit: 6
         slack: array([0., 0., 0.])
        status: 0
       success: True
         upper:  marginals: array([0., 0., 0., 0., 0., 0., 0., 0., 0.])
  residual: array([inf, inf, inf, inf, inf, inf, inf, inf, inf])
             x: array([40., 10., 30.,  0.,  0., 40.,  0., 60.,  0.])

**Análisis de resultados**
El costo mínimo es de \$540.

La cantidad que se transporta de la planta **A** a:  
- **León** es de 40 unidades
- **Guadalajara** es de 10 unidades
- **Morelia** es de 30 unidades


La cantidad que se transporta de la planta **B** a:  
- **León** es de 0 unidades
- **Guadalajara** es de 40 unidades
- **Morelia** es de 0 unidades

La cantidad que se transporta de la planta **C** a:  
- **León** es de 0 unidades
- **Guadalajara** es de 0 unidades
- **Morelia** es de 60 unidades